gnark/gnark-jni/gnark-eip-2537.go

```diff
@@ -173,142 +124,99 @@ eip2537blsG1MultiExp performs a multi-scalar multiplication on multiple G1 point
 		return 1
 	}
 
-	// Convert byte slice to *big.Int and do the initial scalar multiplication
-	scalar := big.NewInt(0)
-	scalar.SetBytes(input[128:160])
-	result := p0.ScalarMultiplication(p0, scalar)
-
-	for i := 1; i < exprCount; i++ {
-		// for each subsequent operation, decode, mul, and add to the result
-		p1, err := g1AffineDecodeInSubGroup(input[i*160 : (i*160)+128])
-		if err != nil {
-			copy(errorBuf, err.Error())
-			return 1
-		}
-
-		scalar = big.NewInt(0)
-		scalar.SetBytes(input[(i*160)+128 : (i+1)*160])
-		p1.ScalarMultiplication(p1, scalar)
-		// add to the result:
-		result = result.Add(result, p1)
-	}
-
-	// marshal the resulting point and encode directly to the output buffer
+	// Store the result of the G1 multi-scalar multiplication into the output buffer
 	return nonMontgomeryMarshalG1(result, javaOutputBuf, errorBuf)
 }
 
-/*
+func _blsG1MultiExpParallel(input []byte, nbTasks int) (*bls12381.G1Affine, error) {
+	var exprCount = len(input) / (EIP2537PreallocateForG1 + EIP2537PreallocateForScalar)
 
-eip2537blsG1MultiExpParallel performs multi-scalar multiplication on multiple G1 points in parallel.
+	// Prepare arrays for points and scalars
+	g1Points := make([]bls12381.G1Affine, exprCount)
+	scalars := make([]fr.Element, exprCount)
 
-- Input:
-	- javaInputBuf: Pointer to a buffer containing a series of G1 point and scalar pairs
-	- javaOutputBuf: Pointer to a buffer where the resulting G1 point will be written
-	- javaErrorBuf: Pointer to a buffer where error messages will be written if an error occurs
-	- cInputLen: Length of the input buffer in bytes
-	- cOutputLen: Length of the output buffer in bytes
-	- cErrorLen: Length of the error buffer in bytes
-	- nbTasks: Number of parallel tasks to use for computation
-- Returns:
-	- zero is returned if successful, result is written to javaOutputBuf
-	- one is returned if there is an error, error message is written to javaErrorBuf
-- Cryptography:
-	- The field elements that comprise the G1 input points must be checked to be canonical.
-	- The scalars are not required to be canonical.
-	- Check that all input points are on the curve and in the correct subgroup.
-- JNI:
-	- javaInputBuf must be at least n*(EIP2537PreallocateForG1 + EIP2537PreallocateForScalar) bytes, where n is the number of point-scalar pairs
-	- javaOutputBuf must be at least EIP2537PreallocateForG1 bytes to safely store the result
+	// Decode points and scalars
+	for i := 0; i < exprCount; i++ {
+		_, err := g1AffineDecodeInSubGroupVal(&g1Points[i], input[i*160:(i*160)+128])
+		if err != nil {
+			return nil, err
+		}
+		scalars[i].SetBytes(input[(i*160)+128 : (i+1)*160])
+	}
 
-*/
-//export eip2537blsG1MultiExpParallel
-func eip2537blsG1MultiExpParallel(javaInputBuf, javaOutputBuf, javaErrorBuf *C.char, cInputLen, cOutputLen, cErrorLen C.int, nbTasks C.int) C.int {
-	inputLen := int(cInputLen)
-	errorLen := int(cOutputLen)
+	if exprCount == 1 {
+		var result bls12381.G1Affine
```

Comment:
a couple comments on what/why we are doing for exprCount == 1 and ==2.  Should be obvious to a cryptographer but not as much to maintainers.